### PR TITLE
Use a transaction for get_range instead of Lua

### PR DESCRIPTION
**Pull request by bmerry**
This provides a significant performance improvement (possibly as much as
3x).

### CODE REVIEW

File: katsdptelstate/aio/redis.py [hunk 1]
```diff
@@ -68,6 +80,26 @@ async def __call__(self, client: aioredis.Redis,
...
+    Returns
+    -------
+    value
+        Boundary value
+    include
+        Whether the boundary value is inclusive (false for ``-`` and ``+``).
+    """
+    if packed_timestamp in {b'-', b'+'}:
+        return packed_timestamp, False
+    elif packed_timestamp[:1] == b'[':
```
Comment (ludwigschwardt) [inline katsdptelstate/aio/redis.py:95]: I like `[:1]` instead of `[0]` - handy trick that.

**Comment by bmerry** [comment]
> What is the difference between a transaction and a script? Why not do everything via transactions?

Scripts are more flexible because they can do things like make a change, then query something, then make another change dependent on that query, and have it all be atomic and avoid round trips to the server. So for example, the script that's getting replaced would first check if the key exists, and skip the other steps if not. With the transaction, we do all the steps unconditionally.

One reason I went with scripts to start with is that the script can be shared between the sync and async interfaces. I've now had to duplicate the logic (with changes, because aioredis and redis don't have quite the same interfaces to either transactions or Z[REV]RANGEBYLEX).

File: katsdptelstate/aio/redis.py (hunk 1 above)
Comment (bmerry) [reply katsdptelstate/aio/redis.py:95]: Also necessary for byte strings in Python 3 if you want a byte string back: `b'xy'[0]` is `120`.

**Comment by bmerry** [comment]
> This provides a significant performance improvement (possibly as much as 3x).

I should clarify that it improves performance on the Redis server, in terms of how long it blocks other clients. It'll potentially increase load on the clients, because they now have to stitch together the results of the two queries.

File: katsdptelstate/redis.py [hunk 2]
```diff
@@ -199,10 +200,22 @@ def get_range(self, key: bytes, start_time: float, end_time: float,
...
+                    pipe.zrangebylex(key, packed_st, packed_et)
+                # raise_on_error annotates the exception message, which breaks
+                # the _handle_wrongtype detection.
+                ret_vals = pipe.execute(raise_on_error=False)
+            for item in ret_vals:
+                if isinstance(item, redis.ResponseError):
+                    raise item
+        if not ret_vals[0]:
+            return None      # Key does not exist
+        print(ret_vals)
```
Comment (sratcliffe) [inline katsdptelstate/redis.py:217]: I assume this will go :)

**Comment by sratcliffe** [comment]
Looks like this at least brings performance back in line with the pre-Lua days - in other words a significant performance improvement.

File: katsdptelstate/redis.py (hunk 2 above)
Comment (bmerry) [reply katsdptelstate/redis.py]: Oops. Fixed.

**Comment by bmerry** [comment]
> Looks like this at least brings performance back in line with the pre-Lua days - in other words a significant performance improvement.

I'm hoping it will also perform a little better than the pre-Lua version for the case of many small queries, because there is now only one round trip to the server, instead of up to 3.

**Pull request merged**